### PR TITLE
Rapid generator for schema-value pairs

### PR DESCRIPTION
**Pull request by t0yv0**
Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/1790 by building a rapid generator for schemas and associated values.

Large-ish problem 1: I do not have it figured out how to test unknown values. TF literals as unknown values are forbidden and do not make sense. We might need a helper resource so that testing unknown values generates references to an output of the helper resource. This is logged for future work.

Large-ish problem 2: iteration is pretty slow (x-proc). Normal n=100 rapid tests can take up to 10min. Could try batching so several resources are tried in one shot say 100 resources. 

Large-ish problem 3: I'm not sure if no-op Update and Create implementations are acceptable. There is something to testing Computed attributes where provider has to set values. Possibly Update also needs to set values? Possibly not.

Small problems:

- [x] Using TF JSON syntax didn't handle null/empty correctly; that is now discarded, using actual HCL syntax
- [x] TF representations are difficult to visualize in failing tests and difficult to assert against
- [x] Lots of lost-in-translation papercuts possible between representations (cty.Value, resource.PropertyValue, tftypes.Value)
- [x] this requires a change to providertest to abstract from testing.T so we can pass rapid.T
- [x] it's very hard to disable annoying TF logging, using env vars for now 

We are starting to find bugs and discrepancies from this work:

- https://github.com/pulumi/pulumi-terraform-bridge/issues/1856 panic corner-case
- https://github.com/pulumi/pulumi-terraform-bridge/issues/1852 need to InternalValidate
- https://github.com/pulumi/pulumi-terraform-bridge/issues/1828

Future work:

- #1856 
- #1857 
- #1858 
- #1859 
- #1860 
- #1861 
- #1862 
- #1863 
- #1864 
- #1865 
- #1866 
- #1867

### CODE REVIEW

**Comment by t0yv0** [comment]
@VenelinMartinov @iwahbe

**Comment by iwahbe** [comment]
> Building #1790
> 
> There's a large number of problems, both small and large, so it's not really finding any real bugs yet.
> 
> Large-ish problem 1: I do not have it figured out how to test unknown values. TF literals as unknown values are forbidden and do not make sense. We might need a helper resource so that testing unknown values generates references to an output of the helper resource.

I'm assuming that this isn't relevant:

https://github.com/pulumi/pulumi-terraform-bridge/blob/cff9ca4eeedebfbd9cb56cdcb9c3bc27b7ad757a/pkg/tfbridge/schema.go#L1001-L1025

> Large-ish problem 2: iteration is pretty slow (x-proc). Normal n=100 rapid tests can take up to 10min. Could try batching so several resources are tried in one shot say 100 resources.

That's painful. Instead of batching, are we missing calls to `t.Parallel()`? In theory, Go should handle batching for us. Do we know what the bottleneck is. 100 tests in 10min indicate 6 seconds per test, which seems high.

> Large-ish problem 3: I'm not sure if no-op Update and Create implementations are acceptable. There is something to testing Computed attributes where provider has to set values. Possibly Update also needs to set values? Possibly not.

This seems like a good problem to address in the next PR.

**Comment by t0yv0** [comment]
6 seconds per test is about right, why would you expect it to be faster? it's shelling out to tf and pulumi CLI and setting up folders and such. 

Looks like rapid does not support parallel exploration of the space. At least I didn't find it. We'd definitely want to batch evaluations over a map of resource instances and shrink it to the failing one.

I've got some simple diff x-tests passing but still iterating to add more of different kinds of generators. Already yielding first very interesting bugs in the tracker.

File: pkg/tests/cross-tests/rapid_tv_gen.go [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+				for f, fg := range fieldGenerators {
+					fv := tvg.drawValue(t, fg, f)
+					fields[f] = fv
+				}
+				return tftypes.NewValue(objType, fields)
+			})
+		} else {
+			objGen = rapid.Just(tftypes.NewValue(objType, map[string]tftypes.Value{}))
+		}
+		// for k, v := range fieldSchemas {
```
Comment (VenelinMartinov) [inline pkg/tests/cross-tests/rapid_tv_gen.go:185]: commented out code, perhaps unintentional?

File: pkg/tests/cross-tests/tf_driver.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	RawPlan  any
+}
+
+func newTfDriver(t T, dir, providerName, resName string, res *schema.Resource) *tfDriver {
+	// Did not find a less intrusive way to disable annoying logging:
+	os.Setenv("TF_LOG_PROVIDER", "off")
+	os.Setenv("TF_LOG_SDK", "off")
+	os.Setenv("TF_LOG_SDK_PROTO", "off")
+
+	// res.CustomizeDiff = func(
```
Comment (VenelinMartinov) [inline pkg/tests/cross-tests/tf_driver.go:60]: commented out code

File: pkg/tests/cross-tests/diff_check.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type diffTestCase struct {
+	// Schema for the resource to test diffing on.
+	Resource *schema.Resource
+
+	// Two resource configurations. The representation assumes JSON Configuration Syntax
```
Comment (VenelinMartinov) [inline pkg/tests/cross-tests/diff_check.go:36]: This is no longer accurate, right? We now write HCL.

File: pkg/tests/cross-tests/tfwrite.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		sch := schemas[key]
+		value, ok := values[key]
+		if !ok {
+			continue
+		}
+		switch elem := sch.Elem.(type) {
+		case *schema.Resource: // TODO sch.ConfigMode
```
Comment (VenelinMartinov) [inline pkg/tests/cross-tests/tfwrite.go:54]: should we add a panic TODO for non-matching ConfigMode?

**Review by VenelinMartinov — APPROVED** [review]
Thanks for pushing this!

File: pkg/tests/cross-tests/rapid_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		outerT.Skip("TODO - we do not currently pass all cases; using this as an exploration tool")
+	}
+	outerT.Parallel()
+
+	log.SetOutput(io.Discard)
+	tvg := &tvGen{}
+
+	rapid.Check(outerT, func(t *rapid.T) {
+		outerT.Logf("Iterating..")
+		tv := tvg.GenBlock(3).Draw(t, "tv")
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply pkg/tests/cross-tests/rapid_test.go:42]: Oh, the depth shouldn't be fixed but should enable shrinking. I'll look into it.

File: pkg/tests/cross-tests/diff_check.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	// See	https://developer.hashicorp.com/terraform/language/syntax/json
+	//
+	// This also accepts tftypes.Value encoded data.
+	Config1, Config2 any
+
+	// Optional object type for the resource.
+	ObjectType *tftypes.Object
+
+	// Bypass interacting with the bridged Pulumi provider.
+	SkipPulumi bool
```
Comment (iwahbe) [inline pkg/tests/cross-tests/diff_check.go:50]: Why would we want to set this?

File: pkg/tests/cross-tests/pu_driver.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+}
+
+func (pd *pulumiDriver) writeYAML(t T, workdir string, tfConfig any) {
+	res := pd.shimProvider.ResourcesMap().Get(pd.tfResourceName)
+	schema := res.Schema()
+	pConfig, err := pd.convertConfigToPulumi(schema, nil, pd.objectType, tfConfig)
+	require.NoErrorf(t, err, "convertConfigToPulumi failed")
+
+	// Not testing secrets yet, but schema secrets may be set by convertConfigToPulumi.
+	pConfig = propertyvalue.RemoveSecrets(resource.NewObjectProperty(pConfig)).ObjectValue()
```
Comment (iwahbe) [inline pkg/tests/cross-tests/pu_driver.go:103]: Can we add a TODO with an associated issue here?

File: pkg/tests/cross-tests/rapid_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tests/cross-tests/rapid_test.go:42]: I've had success with rapid by interpreting the 3 as a depth cap (but not an explicit depth), setting the depth cap to 5 and then letting rapid optimize from there.

File: pkg/tests/cross-tests/rapid_tv_gen.go [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+	typ       tftypes.Object
+	valueGen  *rapid.Generator[tftypes.Value]
+}
+
+type schemaT func(schema.Schema) schema.Schema
+
+type attrKind int
+
+const (
+	optionalAttr attrKind = iota + 1
```
Comment (iwahbe) [inline pkg/tests/cross-tests/rapid_tv_gen.go:47]: Why the `iota + 1`? If `0` is invalid then we should add a `invalidAttr = 0` to make that explicit.

File: pkg/tests/cross-tests/rapid_tv_gen.go [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+			tvg.GenListAttr(maxDepth-1),
+			tvg.GenSetAttr(maxDepth-1),
+		)
+	}
+	return rapid.OneOf(opts...)
+}
+
+func (tvg *tvGen) GenMapAttr(maxDepth int) *rapid.Generator[tv] {
+	ge := rapid.Custom[tv](func(t *rapid.T) tv {
+		inner := tvg.GenAttr(maxDepth).Draw(t, "attrGen")
```
Comment (iwahbe) [inline pkg/tests/cross-tests/rapid_tv_gen.go:90]: Does this need to accommodate https://github.com/hashicorp/terraform-plugin-sdk/issues/62?

File: pkg/tests/cross-tests/rapid_tv_gen.go [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+		listWrapType := tftypes.List{ElementType: inner.typ}
+		listWrap := func(vs []tftypes.Value) tftypes.Value {
+			return tftypes.NewValue(listWrapType, vs)
+		}
+		vg := rapid.Map(rapid.SliceOfN(inner.valueGen, 0, 3), listWrap)
+		return tv{
+			schema: schema.Schema{
+				// TODO get creative with the hash function
```
Comment (iwahbe) [inline pkg/tests/cross-tests/rapid_tv_gen.go:121]: Let's make sure there is an issue associated with this TODO.

**Review by iwahbe — APPROVED** [review]
I'd love to check this in with an example of how to use the rapid generator. I want to make sure people can discover how to use these.

File: pkg/tests/cross-tests/rapid_tv_gen.go (hunk 9 above)
Comment (t0yv0) [reply pkg/tests/cross-tests/rapid_tv_gen.go]: Yes. Done. It was found after this PR.

**Comment by t0yv0** [comment]
@iwahbe the example is https://github.com/pulumi/pulumi-terraform-bridge/blob/t0yv0%2Ftv-generators/pkg/tests/cross-tests/rapid_test.go#L30 although it does not quite yet pass. Once you have a test that you can parameterize by "any schema and value of that schema" you can hook it up to rapid to start tearing through.

**Pull request merged**